### PR TITLE
Move Chat Completions to generated layer

### PR DESCRIPTION
**Pull request by joseharriaga**

### CODE REVIEW

File: .dotnet/src/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -3,4 +3,4 @@
 
 using System.Runtime.CompilerServices;
 
-// [assembly: InternalsVisibleTo("OpenAI.Tests")]
+// [assembly: InternalsVisibleTo("OpenAI.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b197326f2e5bfe2e2a49eb2a05bee871c55cc894325b3775159732ad816c4f304916e7f154295486f8ccabefa3c19b059d51cd19987cc2d31a3195d6203ad0948662f51cc61cc3eb535fc852dfe5159318c734b163f7d1387f1112e1ffe10f83aae7b809c4e36cf2025da5d1aed6b67e1556883d8778eeb63131c029555166de")]
```
Comment (annelo-msft) [inline .dotnet/src/Properties/AssemblyInfo.cs:6]: Not super important, but this isn't needed if we don't plan on signing the tests project.

File: .dotnet/tests/OpenAI.Tests.csproj [hunk 2]
```diff
@@ -4,6 +4,8 @@
     <!--Ignore XML doc comments on test types and members-->
     <NoWarn>$(NoWarn);CS1591</NoWarn>
     <LangVersion>latest</LangVersion>
+    <!--<SignAssembly>True</SignAssembly>
```
Comment (annelo-msft) [inline .dotnet/tests/OpenAI.Tests.csproj:7]: These are just comments and can be removed.

File: .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs [hunk 3]
```diff
@@ -12,10 +12,10 @@ public void Sample01_SimpleChat()
         {
             ChatClient client = new("gpt-3.5-turbo", Environment.GetEnvironmentVariable("OPENAI_API_KEY"));
 
-            ChatCompletion chatCompletion = client.CompleteChat("How does AI work? Explain it in simple terms.");
+            ChatCompletion chatCompletion = client.CompleteChat([new UserChatMessage("How does AI work? Explain it in simple terms.")]);
```
Comment (trrwilson) [inline .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs:15]: Not a thing for this PR -- this is where I'm curious if we could safely add an implicit operator to treat strings as user-mode messages with simple string content, enabling this to just be `client.CompleteChat(["How does AI work?"]);`

File: .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs [hunk 4]
```diff
@@ -12,10 +12,10 @@ public void Sample01_SimpleChat()
...
 
             Console.WriteLine($"[ASSISTANT]:");
-            Console.WriteLine($"{chatCompletion.Content}");
+            Console.WriteLine($"{chatCompletion.Content[0].Text}");
```
Comment (trrwilson) [inline .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs:18]: Also not for this PR -- I wonder if could use our own custom type (`MessageContentPartList : IReadOnlyList<MessageContentPart>`) such that we could override `ToString()` to be friendly to happy-path `Console.WriteLine(chatCompletion.Content);`

File: .dotnet/tests/Samples/Chat/Sample03_FunctionCalling.cs [hunk 5]
```diff
@@ -64,11 +62,11 @@ public void Sample03_FunctionCalling()
             ChatClient client = new("gpt-3.5-turbo", Environment.GetEnvironmentVariable("OPENAI_API_KEY"));
 
             #region
-            List<ChatRequestMessage> messages = [
-                new ChatRequestSystemMessage(
+            List<ChatMessage> messages = [
+                new SystemChatMessage(
```
Comment (trrwilson) [inline .dotnet/tests/Samples/Chat/Sample03_FunctionCalling.cs:66]: Naming nit (can be deferred for later): I think we want to keep the role words and "Message" together given they're a bound/domain concept -- so `ChatSystemMessage` or something "creative" like `SystemMessageForChat`

Comment (trrwilson) [reply .dotnet/tests/Samples/Chat/Sample03_FunctionCalling.cs:66]: (I'd love if we could just drop "chat" from a lot of these types)

File: .dotnet/tests/Samples/Chat/Sample05_ChatWithVision.cs [hunk 6]
```diff
@@ -2,24 +2,33 @@
...
-                    "Describe this image for me",
-                    ChatMessageContent.FromImage(imageUri))
+            string imageFilename = "variation_sample_image.png";
+            string imagePath = Path.Combine("Assets", imageFilename);
+            using Stream imageStream = File.OpenRead(imagePath);
+            BinaryData image = BinaryData.FromStream(imageStream);
+
+            List<ChatMessage> messages = [
+                new UserChatMessage(
+                    ChatMessageContentPart.CreateTextMessageContentPart("Please describe the following image."),
```
Comment (trrwilson) [inline .dotnet/tests/Samples/Chat/Sample05_ChatWithVision.cs:24]: given these methods are static on the type, is it viable for us to remove the duplicative naming segments? `ChatMessageContentPart.CreateText` seems clear enough to me for a static factory.

**Review by trrwilson — APPROVED** [review]
None of my comments are blocking -- this is amazing and we should get it merged ASAP!

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs [hunk 7]
```diff
@@ -1,113 +1,103 @@
...
-            Internal.Models.ChatCompletionFunctions internalFunction = new(
-                function.Description,
-                function.FunctionName,
-                CreateInternalFunctionParameters(function.Parameters),
-                serializedAdditionalRawData: null);
-            internalFunctions.Add(internalFunction);
-        }
-        return internalFunctions;
-    }
+    // TODO: What do we do about this?
```
Comment (trrwilson) [inline .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:43]: IMO we should initialize it and always set the "stream usage" to true. It feels exactly like the embeddings encoding format thing -- they couldn't make it the default out of an abundance of caution for breaking changes, but for libraries there's no good reason not to opt in.

File: .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs (hunk 4 above)
Comment (joseharriaga) [reply .dotnet/tests/Samples/Chat/Sample01_SimpleChat.cs:18]: I like this idea!

File: .dotnet/tests/Samples/Chat/Sample03_FunctionCalling.cs (hunk 5 above)
Comment (joseharriaga) [reply .dotnet/tests/Samples/Chat/Sample03_FunctionCalling.cs:66]: I'll add this comment to the APIView.

File: .dotnet/tests/Samples/Chat/Sample05_ChatWithVision.cs (hunk 6 above)
Comment (joseharriaga) [reply .dotnet/tests/Samples/Chat/Sample05_ChatWithVision.cs:24]: I'll add this comment to the APIView.

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs (hunk 7 above)
Comment (joseharriaga) [reply .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:43]: I added this!

**Pull request merged**